File: uri/strings/growing_strings/1141.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

int main() {
    int n;
    
    while (cin >> n && n != 0) {
        vector<string> strs(n);
        
        for (int i = 0; i < n; i++) {
            cin >> strs[i];
        }
        
        sort(strs.begin(), strs.end(), [](const string& a, const string& b) {
            return a.length() < b.length();
        });
        
        vector<int> dp(n, 1);
        int result = 1;
        
        for (int i = 1; i < n; i++) {
            for (int j = i - 1; j >= 0; j--) {
                if (strs[i].length() - strs[j].length() > 1000) break;
                if (strs[i].find(strs[j]) != string::npos) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
            result = max(result, dp[i]);
        }
        
        cout << result << endl;
    }
    
    return 0;
}

File: uri/strings/growing_strings/1141.rs
use std::io;

fn main() {
    let mut input = String::new();
    
    loop {
        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        let n: usize = input.trim().parse().unwrap();
        
        if n == 0 {
            break;
        }
        
        let mut strs = Vec::new();
        for _ in 0..n {
            input.clear();
            io::stdin().read_line(&mut input).unwrap();
            strs.push(input.trim().to_string());
        }
        
        strs.sort_by_key(|s| s.len());
        
        let mut dp = vec![1; n];
        let mut result = 1;
        
        for i in 1..n {
            for j in (0..i).rev() {
                if strs[j].len() + 1000 < strs[i].len() {
                    break;
                }
                if strs[i].contains(&strs[j]) {
                    dp[i] = dp[i].max(dp[j] + 1);
                }
            }
            result = result.max(dp[i]);
        }
        
        println!("{}", result);
    }
}
